### PR TITLE
refactor: Process logger

### PR DESCRIPTION
**Pull request by big-andy-coates**
### Description 

This PR has two changes to the processing logger framework.

1. First commit: The common code used create a `ProcessingLogger` is moved into `KsqlQueryBuilder`.  This makes the calling code cleaner and, more importantly, makes it easier for tests to mock the processing logger, (making it easier for people will test its use).
2. Second commit: The `ProcessingLogger.error` method is changed to accept an object, not a function. Again, the main reason for this is to make it easier for tests to `verify` what is being logged in unit tests.  (I also moved some classes so that all the processing logger related stuff is in the same package, rather than `util` package.).
3. Third commit: fixes up remaining tests to use `KsqlQueryBuilder.getProcessingLogger` rather than `KsqlQueryBuilder.getProcessingLogContext`.  This simplifies the tests.

I recommend reviewing each commit separately.

While working on this I noticed the timestamp extraction was not actually setting an error on the message being logged!

### Testing done 

usual.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/timestamp/LoggingTimestampExtractor.java [hunk 1]
```diff
@@ -82,26 +78,4 @@ public long extract(final GenericRow row) {
...
-      if (config.getBoolean(ProcessingLogConfig.INCLUDE_ROWS)) {
-        error.put(ProcessingLogMessageSchema.RECORD_PROCESSING_ERROR_FIELD_RECORD, row.toString());
-      }
-
-      final List<String> cause = ErrorMessageUtil.getErrorMessages(e);
-      cause.remove(0);
-      error.put(ProcessingLogMessageSchema.RECORD_PROCESSING_ERROR_FIELD_CAUSE, cause);
-
-      return new SchemaAndValue(ProcessingLogMessageSchema.PROCESSING_LOG_SCHEMA, message);
-    };
```
Comment (big-andy-coates) [inline ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/timestamp/LoggingTimestampExtractor.java]: bug: note how `error` is never set in the returned `message`!

File: ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/timestamp/LoggingTimestampExtractor.java [hunk 2]
```diff
@@ -70,7 +65,8 @@ public long extract(final GenericRow row) {
     try {
       return delegate.extract(row);
     } catch (final Exception e) {
-      logger.error(timestampExtractErroMsg(e, row.toString()));
+      logger.error(RecordProcessingError.recordProcessingError(
```
Comment (rodesai) [inline ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/timestamp/LoggingTimestampExtractor.java]: can we add a test case to `LoggingTimestampExtractorTest`?

File: ksqldb-common/src/main/java/io/confluent/ksql/logging/processing/ProcessingLogger.java [hunk 3]
```diff
@@ -15,14 +15,30 @@
...
 
 public interface ProcessingLogger {
+
+  /**
+   * The interface all error message types must implement.
+   */
+  interface ErrorMessage {
+
+    /**
+     * Called to convert the error message into a structured message.
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksqldb-common/src/main/java/io/confluent/ksql/logging/processing/ProcessingLogger.java]: done.

File: ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/timestamp/LoggingTimestampExtractor.java (hunk 2 above)
Comment (big-andy-coates) [reply ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/timestamp/LoggingTimestampExtractor.java]: Certainly can - as its now easy with this test!

**Pull request merged**